diff -up libtirpc-0.2.1/src/rpc_generic.c.orig libtirpc-0.2.1/src/rpc_generic.c
--- libtirpc-0.2.1/src/rpc_generic.c.orig	2016-12-07 17:12:32.168758752 -0500
+++ libtirpc-0.2.1/src/rpc_generic.c	2016-12-07 17:13:35.918037533 -0500
@@ -606,6 +606,7 @@ __rpc_taddr2uaddr_af(int af, const struc
 	struct sockaddr_in6 *sin6;
 	char namebuf6[INET6_ADDRSTRLEN];
 #endif
+	int path_len;
 	u_int16_t port;
 
 	if (nbuf->len <= 0)
@@ -636,13 +637,12 @@ __rpc_taddr2uaddr_af(int af, const struc
 #endif
 	case AF_LOCAL:
 		sun = nbuf->buf;
-		/*	if (asprintf(&ret, "%.*s", (int)(sun->sun_len -
-		    offsetof(struct sockaddr_un, sun_path)),
-		    sun->sun_path) < 0)*/
-		if (asprintf(&ret, "%.*s", (int)(sizeof(*sun) -
-						 offsetof(struct sockaddr_un, sun_path)),
-			     sun->sun_path) < 0)
 
+		path_len = nbuf->len - offsetof(struct sockaddr_un, sun_path);
+		if (path_len < 0)
+			return NULL;
+
+		if (asprintf(&ret, "%.*s", path_len, sun->sun_path) < 0)
 			return (NULL);
 		break;
 	default:
