diff -up libtirpc-0.2.1/src/svc_dg.c.orig libtirpc-0.2.1/src/svc_dg.c
--- libtirpc-0.2.1/src/svc_dg.c.orig	2013-04-18 13:00:52.570988000 -0400
+++ libtirpc-0.2.1/src/svc_dg.c	2013-04-18 13:03:10.519071000 -0400
@@ -284,7 +284,6 @@ svc_dg_getargs(xprt, xdr_args, args_ptr)
 {
 	if (! SVCAUTH_UNWRAP(xprt->xp_auth, &(su_data(xprt)->su_xdrs),
 			     xdr_args, args_ptr)) {
-		(void)svc_freeargs(xprt, xdr_args, args_ptr);
 		return FALSE;
 	}
 	return TRUE;
