diff -up libtirpc-0.2.1/src/svc_vc.c.orig libtirpc-0.2.1/src/svc_vc.c
--- libtirpc-0.2.1/src/svc_vc.c.orig	2017-11-10 09:10:51.588840720 -0500
+++ libtirpc-0.2.1/src/svc_vc.c	2017-11-10 09:11:11.389237667 -0500
@@ -567,20 +567,19 @@ write_vc(xprtp, buf, len)
 				cd->strm_stat = XPRT_DIED;
 				return (-1);
 			}
-			if (cd->nonblock && i != cnt) {
-				/*
-				 * For non-blocking connections, do not
-				 * take more than 2 seconds writing the
-				 * data out.
-				 *
-				 * XXX 2 is an arbitrary amount.
-				 */
-				gettimeofday(&tv1, NULL);
-				if (tv1.tv_sec - tv0.tv_sec >= 2) {
-					cd->strm_stat = XPRT_DIED;
-					return (-1);
-				}
+			/*
+			 * For non-blocking connections, do not
+			 * take more than 2 seconds writing the
+			 * data out.
+			 *
+			 * XXX 2 is an arbitrary amount.
+			 */
+			gettimeofday(&tv1, NULL);
+			if (tv1.tv_sec - tv0.tv_sec >= 2) {
+				cd->strm_stat = XPRT_DIED;
+				return (-1);
 			}
+			i = 0; /* Don't change buf and cnt */
 		}
 	}
 
