diff -up libtirpc-0.2.1/src/svc_vc.c.orig libtirpc-0.2.1/src/svc_vc.c
--- libtirpc-0.2.1/src/svc_vc.c.orig	2015-11-13 10:43:24.057135642 -0500
+++ libtirpc-0.2.1/src/svc_vc.c	2015-11-13 10:47:27.379451506 -0500
@@ -395,6 +395,12 @@ svc_vc_destroy(xprt)
 	__svc_vc_dodestroy(xprt);
 }
 
+static bool_t
+__svc_rendezvous_socket(xprt)
+	SVCXPRT *xprt;
+{
+	return (xprt->xp_ops->xp_recv == rendezvous_request);
+}
 static void
 __svc_vc_dodestroy(xprt)
 	SVCXPRT *xprt;
@@ -406,7 +412,7 @@ __svc_vc_dodestroy(xprt)
 
 	if (xprt->xp_fd != RPC_ANYFD)
 		(void)close(xprt->xp_fd);
-	if (xprt->xp_port != 0) {
+	if (__svc_rendezvous_socket(xprt)) {
 		/* a rendezvouser socket */
 		r = (struct cf_rendezvous *)xprt->xp_p1;
 		mem_free(r, sizeof (struct cf_rendezvous));
